clouddriver-google/src/main/groovy/com/netflix/spinnaker/clouddriver/google/provider/agent/GoogleServerGroupCachingAgent.groovy

```diff
@@ -55,70 +61,196 @@ class GoogleServerGroupCachingAgent extends AbstractGoogleCachingAgent {
     BatchRequest autoscalerRequest = buildBatchRequest()
 
-    List<String> zones = compute.regions().get(project, region).execute().getZones().collect { Utils.getLocalName(it) }
     zones?.each { String zone ->
-      MIGSCallback migsCallback = new MIGSCallback(serverGroups: serverGroups,
-                                                   zone: zone,
-                                                   instanceGroupsRequest: instanceGroupsRequest,
-                                                   autoscalerRequest: autoscalerRequest)
-      compute.instanceGroupManagers().list(project, zone).queue(migsRequest, migsCallback)
+      InstanceGroupManagerListCallback igmlCallback = new InstanceGroupManagerListCallback(
+          serverGroups: serverGroups,
+          zone: zone,
+          instanceGroupsRequest: instanceGroupsRequest,
+          autoscalerRequest: autoscalerRequest)
+      if (onDemandServerGroupName) {
+        igmlCallback.onDemandServerGroupName = onDemandServerGroupName
+      }
+      compute.instanceGroupManagers().list(project, zone).queue(igmlRequest, igmlCallback)
     }
-    executeIfRequestsAreQueued(migsRequest)
+    executeIfRequestsAreQueued(igmlRequest)
     executeIfRequestsAreQueued(instanceGroupsRequest)
     executeIfRequestsAreQueued(autoscalerRequest)
 
     serverGroups
   }
 
-  CacheResult buildCacheResult(ProviderCache _, List<GoogleServerGroup2> serverGroups) {
+  @Override
+  boolean handles(OnDemandAgent.OnDemandType type, String cloudProvider) {
+    type == OnDemandAgent.OnDemandType.ServerGroup && cloudProvider == GoogleCloudProvider.GCE
+  }
+
+  @Override
+  OnDemandAgent.OnDemandResult handle(ProviderCache providerCache, Map<String, ? extends Object> data) {
+    if (!data.containsKey("serverGroupName") ||
+        data.account != accountName ||
+        data.region != region ||
+        data.zone.empty) {
+      return null
+    }
+
+    GoogleServerGroup2 serverGroup = metricsSupport.readData {
+      getServerGroup(data.serverGroupName as String, data.zone as String)
+    }
+
+    def cacheResultBuilder = new CacheResultBuilder(startTime: Long.MAX_VALUE)
+    CacheResult result = metricsSupport.transformData {
+      buildCacheResult(cacheResultBuilder, [serverGroup])
+    }
+
+    if (result.cacheResults.values().flatten().empty) {
+      // Avoid writing an empty onDemand cache record (instead delete any that may have previously existed).
+      providerCache.evictDeletedItems(ON_DEMAND.ns, [Keys.getServerGroupKey(serverGroup.name,
+                                                                            accountName,
+                                                                            serverGroup.zone)])
+    } else {
+      metricsSupport.onDemandStore {
+        def cacheData = new DefaultCacheData(
+            Keys.getServerGroupKey(data.serverGroupName as String, accountName, data.zone as String),
+            10 * 60, // ttl is 10 minutes
```

Comment:
I know I'm guilty of the same thing, but we should probably switch this to something like `TimeUnits.SECONDS.fromMinutes(10)`